File: 94_binary_tree_inorder_traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}

# RECRUSIVE SOLUTION (TRIVIAL)
# def inorder_traversal(root)
#   return [] if root.nil?

#   inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right)
# end

# ITERATIVE SOLUTION
def inorder_traversal(root)
  results = []
  stack = []
  stack.push(root)
  currNode = root

  until stack.empty?
    if currNode.nil?
      currNode = stack.pop

      if currNode
        results << currNode.val
        currNode = currNode.right
        stack.push(currNode)
      end
    end

    if currNode
      stack.push(currNode.left)
      currNode = currNode.left
    end
  end

  results
end

File: 94_binary_tree_inorder_traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
  let results = [];
  let stack = [root];

  let currNode = root;

  while (stack.length > 0) {
    while (currNode !== null) {
      stack.push(currNode.left);
      currNode = currNode.left;
    }

    currNode = stack.pop();

    if (currNode !== null) {
      results.push(currNode.val);
      stack.push(currNode.right);
      currNode = currNode.right;
    }
  }

  return results;
};